zebra-consensus/src/block/subsidy/funding_streams.rs

```diff
@@ -0,0 +1,113 @@
+//! Funding Streams calculations. - [§7.7][7.7]
+//!
+//! [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+
+use std::convert::TryFrom;
+
+use zebra_chain::{
+    amount::{Amount, Error, NonNegative},
+    block::Height,
+    parameters::{Network, NetworkUpgrade::*},
+};
+
+use crate::{
+    block::subsidy::general::block_subsidy,
+    parameters::subsidy::{
+        FundingStreamReceiver, FUNDING_STREAM_HEIGHT_RANGES, FUNDING_STREAM_RECEIVER_DENOMINATOR,
+        FUNDING_STREAM_RECEIVER_NUMERATORS,
+    },
+};
+
+/// Returns the `fs.Value(height)` for each stream receiver
+/// as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+use std::collections::HashMap;
+pub fn funding_stream_values(
+    height: Height,
+    network: Network,
+) -> Result<HashMap<FundingStreamReceiver, Amount<NonNegative>>, Error> {
+    let canopy_height = Canopy.activation_height(network).unwrap();
+    let mut results = HashMap::new();
+
+    if height >= canopy_height {
+        let range = FUNDING_STREAM_HEIGHT_RANGES.get(&network).unwrap();
+        if range.contains(&height) {
+            for (&receiver, &numerator) in FUNDING_STREAM_RECEIVER_NUMERATORS.iter() {
+                let amount_value = (block_subsidy(height, network)?.number() as f64
+                    * (numerator as f64 / FUNDING_STREAM_RECEIVER_DENOMINATOR as f64))
+                    .floor();
```

Comment:
We can't use `f64` in consensus-critical calculations, because it doesn't guarantee integer precision.
And it doesn't handle overflow the same way that integers do.

`Mul<u64> for Amount` and `Div<u64> for Amount<C>` already check for overflow:
https://github.com/ZcashFoundation/zebra/blob/bdf5f70557374fbec2f21c96aac23caa267f07e0/zebra-chain/src/amount.rs#L314-L330

https://github.com/ZcashFoundation/zebra/blob/bdf5f70557374fbec2f21c96aac23caa267f07e0/zebra-chain/src/amount.rs#L345-L354

Reply:
Thanks, i was able to make the calculation using `Amount`s in https://github.com/ZcashFoundation/zebra/pull/3017/commits/7d7ac7fab9932281eafdceae600adc131cb56b87

Thanks!